### PR TITLE
feat(2fa): Add Deleting 2FA

### PR DESCRIPTION
**Pull request by ice-tychon**
### What does this PR do?
This PR adds ability to delete previously added 2FA options

### Changes brought by this PR
- The main changes are in
 * ion_identity_client: added code to delete 2FA
 * features/protect_account: added UI for deleting 2FA options and connect it to ion_identity_client

### Design
https://www.figma.com/design/3TnZngoFeklOVC8TexlR1D/%F0%9F%A7%8A-ICE-Cloud?node-id=44043-159152&m=dev

### Screenshots
CleanShot 2024-11-07 at 15:52:35
CleanShot 2024-11-07 at 15:53:05
CleanShot 2024-11-07 at 15:53:15
CleanShot 2024-11-07 at 15:53:23
CleanShot 2024-11-07 at 15:53:31

### CODE REVIEW

File: lib/app/features/protect_account/components/delete_twofa_input_step.dart [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  void _onConfirm(
+    WidgetRef ref,
+    BuildContext context,
+    Map<TwoFaType, TextEditingController> controllers,
+  ) {
+    ref.read(deleteTwoFANotifierProvider.notifier).deleteTwoFa(
+      const TwoFAType.authenticator(),
+      [
+        for (final e in controllers.entries) TwoFaTypeAdapter(e.key, e.value.text).twoFAType,
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/components/delete_twofa_input_step.dart]: Done.

File: lib/app/features/protect_account/components/delete_twofa_input_step.dart [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            Button(
+              mainAxisSize: MainAxisSize.max,
+              label: Text(context.i18n.button_confirm),
+              onPressed: () {
+                final isFormValid = formKey.value.currentState!.validate();
+                if (!isFormValid) {
+                  return;
+                }
+
+                _onConfirm(ref, context, controllers);
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/components/delete_twofa_input_step.dart]: Done.

File: lib/app/features/protect_account/components/delete_twofa_input_step.dart [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  final List<TwoFaType> twoFaTypes;
+  final VoidCallback onDeleteSuccess;
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final isRequesting = ref.watch(requestTwoFaCodeNotifierProvider).isLoading;
+
+    final formKey = useRef(GlobalKey<FormState>());
+
+    final controllers = {
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/components/delete_twofa_input_step.dart]: Done.

File: lib/app/features/protect_account/components/delete_twofa_input_step.dart [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    required this.onDeleteSuccess,
+    super.key,
+  });
+
+  final List<TwoFaType> twoFaTypes;
+  final VoidCallback onDeleteSuccess;
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final isRequesting = ref.watch(requestTwoFaCodeNotifierProvider).isLoading;
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/components/delete_twofa_input_step.dart]: Done.

File: lib/app/features/auth/views/pages/twofa_codes/twofa_code_input.dart [hunk 5]
```diff
@@ -71,13 +83,15 @@ class SendButton extends HookWidget {
...
-              isSent.value = true;
-              startCountdown();
-            },
-            label: isSent.value ? context.i18n.button_retry : context.i18n.button_send,
-          );
+        : isSending
+            ? IONLoadingIndicator(type: isLightTheme ? IndicatorType.dark : IndicatorType.light)
+            : TextInputTextButton(
+                onPressed: () {
+                  onRequestCode?.call();
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/auth/views/pages/twofa_codes/twofa_code_input.dart]: Done.

File: lib/app/features/protect_account/phone/views/pages/setup_phone/phone_setup_input_page.dart [hunk 6]
```diff
@@ -66,14 +70,25 @@ class PhoneSetupInputPage extends HookConsumerWidget {
...
-                  onPressed: () {
-                    if (formKey.value.currentState?.validate() ?? false) {
-                      final fullPhoneNumber = '${country.iddCode}${phoneController.text.trim()}';
+                  onPressed: () async {
+                    final isFormValid = formKey.value.currentState?.validate() ?? false;
+                    if (!isFormValid) {
+                      return;
+                    }
+
+                    final fullPhoneNumber = '${country.iddCode}${phoneController.text.trim()}';
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/phone/views/pages/setup_phone/phone_setup_input_page.dart:79]: Done.

File: lib/app/features/protect_account/email/providers/linked_phone_provider.dart [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+      ) ??
+      0;
+
+  final phone = phones[phoneIndex];
+  if (phone.length <= 6) {
+    return phone;
+  }
+  final start = phone.substring(0, 5);
+  final end = phone.substring(phone.length - 2);
+  return '$start***$end';
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/email/providers/linked_phone_provider.dart:36]: Done.

File: lib/app/features/protect_account/email/providers/linked_email_provider.dart [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        userDetails.twoFaOptions
+                ?.firstWhereOrNull(startsWith(const TwoFAType.email().option))
+                ?.split(':')
+                .firstOrNull ??
+            '0',
+      ) ??
+      0;
+
+  final email = emails[emailIndex];
+  final atIndex = email.indexOf('@');
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/features/protect_account/email/providers/linked_email_provider.dart]: Done.

File: lib/app/services/ion_identity/ion_identity_client_provider.dart [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import 'package:ion_identity_client/ion_identity.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'ion_identity_client_provider.g.dart';
+
+@riverpod
+Future<Raw<IONIdentityClient>> ionIdentityClient(Ref ref) async {
+  final currentUser = ref.watch(currentIdentityKeyNameSelectorProvider);
+  if (currentUser == null) {
+    throw Exception('Current user not found');
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/services/ion_identity/ion_identity_client_provider.dart:15]: Done.

**Review by ice-kreios — DISMISSED** [review]

File: lib/app/utils/future.dart [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+/// Adds extensions to num (ie. int & double) to make creating durations simple:
+///
+/// ```
+/// 200.ms // equivalent to Duration(milliseconds: 200)
+/// 3.seconds // equivalent to Duration(milliseconds: 3000)
+/// 1.5.days // equivalent to Duration(hours: 36)
+///
+/// Copied from flutter_animate package
+/// ```
+extension NumDurationExtensions on num {
```
[the remark this answers is not in the document]
Comment (ice-tychon) [reply lib/app/utils/future.dart]: Fixed in another way.

**Review by ice-alcides — APPROVED** [review]

**Pull request merged**